File: samplesJS/Particles/src/ParticlesSample.cpp
// Required for WinMain
#include <Windows.h>

// Oak Nut include
#include <onut/Anim.h>
#include <onut/ParticleSystem.h>
#include <onut/ParticleSystemManager.h>
#include <onut/Font.h>
#include <onut/Input.h>
#include <onut/onut.h>
#include <onut/Renderer.h>
#include <onut/Settings.h>
#include <onut/SpriteBatch.h>
#include <onut/Timing.h>

void init();
void update();
void render();

OEmitterInstance emitter;
OAnimVector3 position;

// Main
int CALLBACK WinMain(HINSTANCE appInstance, HINSTANCE prevInstance, LPSTR cmdLine, int cmdCount)
{
    oSettings->setGameName("Particles Sample");
    ORun(init, update, render);
}

void init()
{
}

void update()
{
    emitter.setTransform(position);

    if (OInputJustPressed(OKey1))
    {
        OEmitParticles("test.pfx", Vector3{OScreenWf / 3, OScreenHf / 2, 0});
    }
    if (OInputJustPressed(OKey2))
    {
        if (emitter.isPlaying())
        {
            emitter.stop();
        }
        else
        {
            // Animate it back and forth across the screen
            position.play(Vector3{OScreenWf - 100, OScreenHf / 2, 0}, Vector3{100, OScreenHf / 2, 0}, 2.f, OTweenLinear, OPingPongLoop);

            emitter = OEmitParticles("test2.pex", position);
            emitter.setRenderEnabled(false);
        }
    }
}

void render()
{
    // Clear to black
    oRenderer->clear({0, 0, 0, 1});

    oSpriteBatch->begin();

    // Draw info
    auto pFont = OGetFont("font.fnt");
    pFont->draw("Press ^9901^999 to spawn particles from an onut PFX file", {10, 10});
    pFont->draw("Press ^9902^999 to spawn particles from PEX file", {10, 30});

    // It is possible to manually call render an emitter,
    // so we can specify in which order it is renderer manually
    emitter.render();

    pFont->draw("FPS: " + std::to_string(oTiming->getFPS()), {10, 50});

    oSpriteBatch->end();
}

File: samplesJS/Particles/assets/scripts/ParticlesSample.js
var font = getFont("font.fnt");
var particleSystem1 = getParticleSystem("test.pfx");
var particleSystem2 = getParticleSystem("test2.pex");

var emitter = new ParticleEmitter();
var position = new Vector3Anim();

function update()
{
    var screen = Renderer.getResolution();

    emitter.setTransform(position.get());

    if (Input.isJustDown(Key._1))
    {
        emitParticles(particleSystem1, new Vector3(screen.x / 3, screen.y / 2, 0));
    }
    if (Input.isJustDown(Key._2))
    {
        if (emitter.isPlaying())
        {
            emitter.stop();
        }
        else
        {
            // Animate it back and forth across the screen
            position.playSingle(new Vector3(screen.x - 100, screen.y / 2, 0), new Vector3(100, screen.y / 2, 0), 2, Tween.LINEAR, Loop.PING_PONG_LOOP);

            emitter = emitParticles(particleSystem2, position.get());
            emitter.setRenderEnabled(false);
        }
    }
}

function render()
{
    // Clear to black
    Renderer.clear(Color.BLACK);

    SpriteBatch.begin();

    // Draw info
    SpriteBatch.drawText(font, "Press ^9901^999 to spawn particles from an onut PFX file", new Vector2(10, 10));
    SpriteBatch.drawText(font, "Press ^9902^999 to spawn particles from PEX file", new Vector2(10, 30));

    // It is possible to manually call render an emitter,
    // so we can specify in which order it is renderer manually
    emitter.render();

    SpriteBatch.drawText(font, "FPS: " + Timing.getFPS(), new Vector2(10, 50));

    SpriteBatch.end();
}
